### PR TITLE
CWE-016 - Adds MaxRequestLength check for ASP.NET

### PR DESCRIPTION
**Pull request by cldrn**
maxrequestlength

### CODE REVIEW

File: csharp/ql/src/Security Features/CWE-016/ASPNetMaxRequestLength.qhelp [hunk 1]
```diff
@@ -0,0 +1,22 @@
+<!DOCTYPE qhelp PUBLIC
+  "-//Semmle//qhelp//EN"
+  "qhelp.dtd">
+<qhelp>
+
+<overview>
+<p>
+The MaxRequestLength attribute sets the limit for the input stream buffering threshold in KB. Attackers can use large files to cause denial of service attacks.
```
Comment (hvitved) [inline csharp/ql/src/Security Features/CWE-016/ASPNetMaxRequestLength.qhelp]: `MaxRequestLength` -> `<code>MaxRequestLength</code>`

File: csharp/ql/src/Security Features/CWE-016/ASPNetMaxRequestLength.ql [hunk 2]
```diff
@@ -0,0 +1,18 @@
+/**
+ * @name Large maxRequestLength value.
```
Comment (hvitved) [inline csharp/ql/src/Security Features/CWE-016/ASPNetMaxRequestLength.ql]: No period at the end.

File: csharp/ql/src/Security Features/CWE-016/ASPNetMaxRequestLength.ql [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+ * @description maxRequestLength should be set as minimum as possible.
+ * @kind problem
+ * @problem.severity warning
+ * @precision very-high
+ * @id cs/web/debug-binary
```
Comment (hvitved) [inline csharp/ql/src/Security Features/CWE-016/ASPNetMaxRequestLength.ql]: `cs/web/debug-binary` -> `cs/web/max-request-length`

File: csharp/ql/src/Security Features/CWE-016/ASPNetMaxRequestLength.ql [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+ * @tags security
+ *       maintainability
+ *       frameworks/asp.net
+ *       external/cwe/cwe-16
+ */
+
+import csharp
+from XMLAttribute a
+where
+  a.getName().toLowerCase() = "maxrequestlength" and a.getValue() > "4096"
```
Comment (hvitved) [inline csharp/ql/src/Security Features/CWE-016/ASPNetMaxRequestLength.ql]: The test `a.getValue() > "4096"` actually performs a lexicographic string comparison; `a.getValue().toInt() > 4096` is better.

File: csharp/ql/src/Security Features/CWE-016/ASPNetMaxRequestLength.ql [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+select a, "maxRequestLength is set too large: "+a.getValue()+" KB. It is recommended to set it as minimum as possible based on business requirements."
```
Comment (hvitved) [inline csharp/ql/src/Security Features/CWE-016/ASPNetMaxRequestLength.ql]: `select a, "Large 'maxRequestLength' value (" + a.getValue() + ")."`

File: csharp/ql/src/Security Features/CWE-016/ASPNetMaxRequestLength.ql [hunk 6]
```diff
@@ -0,0 +1,18 @@
+/**
+ * @name Large maxRequestLength value.
+ * @description maxRequestLength should be set as minimum as possible.
+ * @kind problem
+ * @problem.severity warning
+ * @precision very-high
```
Comment (hvitved) [inline csharp/ql/src/Security Features/CWE-016/ASPNetMaxRequestLength.ql]: Please remove the precision tag (we will make sure to readd it in a follow-up PR.)

**Comment by cldrn** [comment]
Thanks for the feedback. I've address all your comments. I wrote them while analyzing my first project so I'm still reading up on the documentation. 

Awesome work!

File: csharp/ql/src/Security Features/CWE-016/ASPNetMaxRequestLength.ql [hunk 7]
```diff
@@ -0,0 +1,23 @@
+/**
+ * @name Large 'maxRequestLength' value
+ * @description Setting a large 'maxRequestLength' value may render a web page vulnerable to
```
Comment (jf205) [inline csharp/ql/src/Security Features/CWE-016/ASPNetMaxRequestLength.ql]:
```suggestion
 * @description Setting a large 'maxRequestLength' value may render a webpage vulnerable to
```

File: csharp/ql/src/Security Features/CWE-016/ASPNetMaxRequestLength.qhelp [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+<example>
+
+<p>
+The following example shows the <code>maxRequestLength</code> attribute set to a high value (256 MB) in a <code>Web.config</code> file for ASP.NET:
+</p>
+
+<sample src="Web.config.bad" />
+
+<p>
+Unless such a high value is strictly needed, it is better to set a lower value, for example 4096:
```
Comment (jf205) [inline csharp/ql/src/Security Features/CWE-016/ASPNetMaxRequestLength.qhelp]: Suggest using consistent units (KB) in the examples.

File: csharp/ql/src/Security Features/CWE-016/ASPNetMaxRequestLength.qhelp [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+<overview>
+<p>
+The <code>maxRequestLength</code> attribute sets the limit for the input stream buffering threshold in KB. Attackers can use large files to cause denial-of-service attacks.
+</p>
+
+</overview>
+<recommendation>
+
+<p>
+The recommended value is 4096 but you should try setting it as small as possible according to business requirements.
```
Comment (jf205) [inline csharp/ql/src/Security Features/CWE-016/ASPNetMaxRequestLength.qhelp]: Perhaps change this sentence to:

The recommended value is 4096 _KB_, but you should set it as small as possible according to business requirements.

(Note, include units)

File: change-notes/1.23/analysis-csharp.md [hunk 10]
```diff
@@ -9,6 +9,7 @@ The following changes in version 1.23 affect C# analysis in all applications.
 | **Query**                   | **Tags**  | **Purpose**                                                        |
 |-----------------------------|-----------|--------------------------------------------------------------------|
 | Unsafe year argument for 'DateTime' constructor (`cs/unsafe-year-construction`) | reliability, date-time | Finds incorrect manipulation of `DateTime` values, which could lead to invalid dates. |
+| Large 'maxRequestLength' value (`cs/web/large-max-request-length`) | security, frameworks/asp.net, external/cwe/cwe-16 | Finds `web.config` files with large (greater than 4096) `maxRequestLength` attributes. |
```
Comment (jf205) [inline change-notes/1.23/analysis-csharp.md]:
```suggestion
| Large 'maxRequestLength' value (`cs/web/large-max-request-length`) | security, frameworks/asp.net, external/cwe/cwe-16 | Finds `web.config` files with large (greater than 4096 KB) `maxRequestLength` attributes. |
```

File: csharp/ql/src/Security Features/CWE-016/ASPNetMaxRequestLength.qhelp [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+<sample src="Web.config.bad" />
+
+<p>
+Unless such a high value is strictly needed, it is better to set a lower value, for example 4096:
+</p>
+
+<sample src="Web.config.good" />
+
+</example>
+
```
Comment (jf205) [inline csharp/ql/src/Security Features/CWE-016/ASPNetMaxRequestLength.qhelp]: Is there a reference to some authoritative docs you could include here?

**Comment by cldrn** [comment]
Closed this issue as I removed that local copy and started doing each PR in their own branch. At this point I have deleted the original branch for that PR. I've addressed the comments in there but can we keep working on this one https://github.com/Semmle/ql/pull/2355?

The other fork had a  PR on master branch which became a mess when I added others there by mistake. Hence, why I started from stratch with each PR in their own branch.

:)

**Comment by hvitved** [comment]
Closing in favor of https://github.com/Semmle/ql/pull/2355.

**Pull request closed without merging**